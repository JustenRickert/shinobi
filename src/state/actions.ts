import { IR, pipe } from "../util";
import { GameState, Genin, Task } from "./";

export function assignTaskToShinobi(
  task: Task.T,
  shinobiId: Genin.Id,
  state: GameState
) {
  return pipe<GameState>(
    (state) => ({
      ...state,
      genin: IR.update(
        shinobiId,
        (shinobi) => ({
          ...shinobi,
          behavior: Genin.behaviorAssignedTask(task),
        }),
        state.genin
      ),
    }),
    (state) => ({
      ...state,
      village: {
        ...state.village,
        tasks: IR.remove(task.id, state.village.tasks),
      },
    })
  )(state);
}
