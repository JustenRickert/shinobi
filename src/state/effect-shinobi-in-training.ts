import { clamp, range } from "ramda";
import { Observable, interval, map, merge, timeInterval } from "rxjs";

import { GameEffectAction, GameState, ShinobiInTraining } from ".";
import {
  VILLAGE_SHINOBI_IN_TRAINING_GENERATION_TIMEOUT_BASE,
  VILLAGE_SHINOBI_IN_TRAINING_GRADUATION_TIMEOUT_BASE,
} from "../constant";
import { IR } from "../ir";
import { pipeM } from "../util";

let lastTime = Date.now();

function debugLog(msg: any) {
  console.log("since last: %s", Date.now() - lastTime, {
    graduationCount: msg.graduatingShinobi.ids.length,
    graduatingShinobi: msg.graduatingShinobi,
  });
  lastTime = Date.now();
}

function graduate(
  { timestamp }: { timestamp: number },
  state: GameState
): GameState {
  const graduatingShinobi = IR.filter(
    (shinobi) =>
      timestamp - shinobi.createdAt >
      VILLAGE_SHINOBI_IN_TRAINING_GRADUATION_TIMEOUT_BASE,
    state.shinobiInTraining
  );
  debugLog({
    now: timestamp,
    graduatingShinobi,
  });
  return {
    ...state,
    population: state.population + graduatingShinobi.ids.length,
    lastGraduationAt: timestamp,
    shinobiInTraining: IR.remove(
      graduatingShinobi.ids,
      state.shinobiInTraining
    ),
  };
}

function addShinobiInTraining(shinobis: ShinobiInTraining.T[]) {
  return pipeM<GameState>((state) => ({
    ...state,
    shinobiInTraining: IR.add(shinobis, state.shinobiInTraining),
  }));
}

export function effectShinobiInTrainingGenerationAndGraduation(
  _state$: Observable<GameState>
): GameEffectAction {
  const gradTime = VILLAGE_SHINOBI_IN_TRAINING_GRADUATION_TIMEOUT_BASE;
  const genTime = VILLAGE_SHINOBI_IN_TRAINING_GENERATION_TIMEOUT_BASE;

  const generation$ = interval(
    VILLAGE_SHINOBI_IN_TRAINING_GENERATION_TIMEOUT_BASE
  ).pipe(
    // TODO need to account for the amount of population gained
    timeInterval(),
    map(({ interval }) => {
      const count = Math.round(
        clamp(1, gradTime / genTime, interval / genTime)
      );
      console.log("Adding %s shinobi. %ss passed.", count, interval / 1e3, {});
      const now = Date.now();
      return pipeM(
        (state) => {
          const sinceLastGraduation = now - state.lastGraduationAt;
          console.log("since last graduation: %s", sinceLastGraduation);
          if (sinceLastGraduation > gradTime) {
            return graduate({ timestamp: now }, state);
          }
          return state;
        },
        addShinobiInTraining(
          range(0, count).map((i) =>
            ShinobiInTraining.make({
              createdAt: now - genTime * i,
            })
          )
        )
      );
    })
  );

  return merge(generation$);
}
