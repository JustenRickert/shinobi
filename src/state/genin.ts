import { Genin as Genin, ShinobiInTraining, Task } from ".";
import { deviationFloat, uniqueId } from "../util";

export type Id = `shinobi-${string}`;

export interface BehaviorAssignedTask {
  type: "assigned-task";
  task: Task.T;
  since: number;
}

export interface BehaviorAvailable {
  type: "available";
  since: number;
}

export interface BehaviorIdle {
  type: "idle";
  since: number;
}

export interface BehaviorInjured {
  type: "injured";
  since: number;
}

export type Behavior =
  | BehaviorIdle
  | BehaviorAssignedTask
  | BehaviorInjured
  | BehaviorAvailable;

export function behaviorAssignedTask(task: Task.T): BehaviorAssignedTask {
  return {
    type: "assigned-task",
    task,
    since: Date.now(),
  };
}

export function behaviorIdle(): BehaviorIdle {
  return {
    type: "idle",
    since: Date.now(),
  };
}

export function behaviorAvailable(): BehaviorAvailable {
  return {
    type: "available",
    since: Date.now(),
  };
}

export function behaviorInjured(): BehaviorInjured {
  return {
    type: "injured",
    since: Date.now(),
  };
}

export type Message =
  | {
      type: "turned-idle";
      when: number;
    }
  | {
      type: "task-failed";
      when: number;
      task: Task.T;
    }
  | {
      type: "task-success";
      when: number;
      task: Task.T;
    };

export interface T {
  id: Id;
  name: string;
  level: number;
  behavior: Behavior;
  messages: Message[];
}

export function make(shinobi?: ShinobiInTraining.T): T {
  const level = shinobi?.level ?? 0;
  const name = shinobi?.name;
  return {
    id: `shinobi-${uniqueId()}`,
    name: name ?? "Genin Joe",
    level,
    behavior: behaviorAvailable(),
    messages: [],
  };
}

export function addMessage(message: Message) {
  return (shinobi: Genin.T): Genin.T => ({
    ...shinobi,
    messages: [...shinobi.messages, message].slice(-10),
  });
}

function sigmoid(
  x: number,
  c?: { mid?: number; limit?: number; growth?: number }
) {
  const x0 = c?.mid ?? 0;
  const L = c?.limit ?? 1;
  const k = c?.growth ?? 1;
  return L / (1 + Math.E ** (-k * (x + x0)));
}

export function taskSuccessChance(task: Task.T, shinobi: T) {
  return sigmoid(shinobi.level - task.level, { mid: Math.log(3) });
}

export function taskExperienceGain(task: Task.T, shinobi: T) {
  const overlevel = Math.max(0, shinobi.level - task.level);
  return Math.ceil(
    deviationFloat(5 * (1 + task.level) ** 1.15 * 0.85 ** overlevel, 0.15)
  );
}
