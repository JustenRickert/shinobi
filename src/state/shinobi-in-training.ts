import { uniqueId } from "../util";

export type Id = `shinobi-in-training-${string}`;

export interface T {
  id: Id;
  name: string;
  level: number;
  createdAt: number;
}

export function make(): T {
  return {
    id: `shinobi-in-training-${uniqueId()}`,
    name: "Shinobi John",
    level: 0,
    createdAt: Date.now(),
  };
}

export function cost(shinobi: T) {
  return 5 * 1.25 ** shinobi.level;
}
