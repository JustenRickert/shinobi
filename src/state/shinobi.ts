import { Task } from ".";
import { uniqueId } from "../util";

export type Id = `shinobi-${string}`;

export interface T {
  id: Id;
  name: string;
  level: number;
  assignedTask: null | Task.T;
}

export function make(): T {
  return {
    id: `shinobi-${uniqueId()}`,
    name: "Shinobi John",
    level: 0,
    assignedTask: null,
  };
}

export function cost(shinobi: T) {
  return 5 * 1.25 ** shinobi.level;
}
