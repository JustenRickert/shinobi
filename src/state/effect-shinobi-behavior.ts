import {
  EMPTY,
  Observable,
  distinctUntilChanged,
  distinctUntilKeyChanged,
  map,
  of,
  switchMap,
  timer,
  withLatestFrom,
} from "rxjs";
import { IR, assert, chance, pipe, sample, thru, updateIn } from "../util.ts";
import { GameState, Shinobi, Task } from "./";
import { assignTaskToShinobi } from "./actions.ts";
import {
  SHINOBI_ASSIGNED_TASK_TIMEOUT_BASE,
  SHINOBI_IDLE_TASK_TIMEOUT_BASE,
  SHINOBI_INJURED_TIMEOUT_BASE,
} from "../constant";
import { addMessage } from "./shinobi.ts";

export function makeShinobiBehoviorUpdateStream(
  id: Shinobi.Id,
  state$: Observable<GameState>
) {
  const taskSuccessFn = (task: Task.T) => (state: GameState) => ({
    ...state,
    points: state.points + task.points,
    shinobi: thru(
      state.shinobi,
      IR.update(
        id,
        pipe(
          addMessage({ type: "task-success", when: Date.now(), task }),
          (shinobi) => ({
            ...shinobi,
            behavior: {
              type: "idle",
            } as Shinobi.BehaviorIdle,
          })
        )
      )
    ),
  });

  const taskFailureFn = (task: Task.T) => (state: GameState) => ({
    ...state,
    shinobi: thru(
      state.shinobi,
      IR.update(
        id,
        pipe(
          addMessage({ type: "task-failed", when: Date.now(), task }),
          (shinobi) => ({
            ...shinobi,
            behavior: Shinobi.behaviorInjured(),
          })
        )
      )
    ),
  });

  const assignRandomTask = (shinobiId: Shinobi.Id) => (state: GameState) => {
    const task = sample(IR.list(state.village.tasks));
    return assignTaskToShinobi(task, shinobiId, state);
  };

  const turnIdle = (shinobiId: Shinobi.Id) =>
    pipe<GameState>((state: GameState) => ({
      ...state,
      shinobi: thru(
        state.shinobi,
        IR.update(shinobiId, (shinobi) => ({
          ...shinobi,
          behavior: Shinobi.behaviorIdle(),
        }))
      ),
    }));

  const shinobi$ = state$.pipe(
    map((state) => state.shinobi.record[id]),
    distinctUntilChanged()
  );

  return shinobi$.pipe(
    map((s) => s),
    distinctUntilKeyChanged("behavior", (b1, b2) => b1.type === b2.type),
    switchMap((shinobi) => {
      switch (shinobi.behavior.type) {
        case "injured": {
          return timer(SHINOBI_INJURED_TIMEOUT_BASE).pipe(
            map(() => turnIdle(shinobi.id))
          );
        }
        case "idle": {
          return timer(SHINOBI_IDLE_TASK_TIMEOUT_BASE).pipe(
            switchMap(() =>
              state$.pipe(
                map((state) => Boolean(IR.list(state.village.tasks).length)),
                distinctUntilChanged(),
                switchMap((going) =>
                  going ? of(assignRandomTask(shinobi.id)) : EMPTY
                )
              )
            )
          );
        }
        case "assigned-task":
          return timer(SHINOBI_ASSIGNED_TASK_TIMEOUT_BASE).pipe(
            withLatestFrom(shinobi$),
            map(([, shinobi]) => {
              assert(shinobi.behavior.type === "assigned-task");
              const p = Shinobi.taskSuccessChance(
                shinobi.behavior.task,
                shinobi
              );
              const success = chance(p);
              return success
                ? taskSuccessFn(shinobi.behavior.task)
                : taskFailureFn(shinobi.behavior.task);
            })
          );
      }
    })
  );
}
