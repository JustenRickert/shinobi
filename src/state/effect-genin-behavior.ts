import { lensProp, over, set } from "ramda";
import {
  EMPTY,
  Observable,
  distinctUntilChanged,
  distinctUntilKeyChanged,
  map,
  of,
  switchMap,
  timer,
  withLatestFrom,
} from "rxjs";
import {
  SHINOBI_ASSIGNED_TASK_TIMEOUT_BASE,
  SHINOBI_IDLE_TASK_TIMEOUT_BASE,
  SHINOBI_INJURED_TIMEOUT_BASE,
} from "../constant.ts";
import { IR, assert, chance, pipeM, sample, thru } from "../util.ts";
import { assignTaskToShinobi } from "./actions.ts";
import { addMessage } from "./genin.ts";
import { GameState, Genin, Task } from "./index.tsx";

export function makeShinobiBehoviorUpdateStream(
  id: Genin.Id,
  state$: Observable<GameState>
) {
  const taskSuccessFn = (task: Task.T) =>
    pipeM<GameState>(
      over(lensProp("points"), (p) => p + task.points),
      over(
        lensProp("genin"),
        IR.update(
          id,
          pipeM(
            addMessage({ type: "task-success", when: Date.now(), task }),
            set(lensProp("behavior"), Genin.behaviorIdle())
          )
        )
      )
    );

  const taskFailureFn = (task: Task.T) =>
    pipeM<GameState>(
      over(
        lensProp("genin"),
        IR.update(
          id,
          pipeM(
            addMessage({ type: "task-failed", when: Date.now(), task }),
            set(lensProp("behavior"), Genin.behaviorIdle())
          )
        )
      )
    );

  const assignRandomTask = (state: GameState) => {
    // TODO: Sometimes two shinobi are asked to run this at the same time,
    // wherein only one of them will succeed. Here we return early for that
    // case, which is a bit awkward. It may be better to have a separate effect
    // that spies both ready shinobi and available tasks, and dispatches
    if (!state.village.tasks.ids.length) return;
    const task = sample(IR.list(state.village.tasks));
    return assignTaskToShinobi(task, id, state);
  };

  const turnIdle = pipeM<GameState>((state: GameState) => ({
    ...state,
    genin: thru(
      state.genin,
      IR.update(id, (genin) => ({
        ...genin,
        behavior: Genin.behaviorIdle(),
      }))
    ),
  }));

  const turnAvailable = pipeM<GameState>((state: GameState) => ({
    ...state,
    genin: thru(
      state.genin,
      IR.update(id, (shinobi) => ({
        ...shinobi,
        behavior: Genin.behaviorAvailable(),
      }))
    ),
  }));

  const genin$ = state$.pipe(
    map((state) => state.genin.record[id]),
    distinctUntilChanged()
  );

  const shinobiBehaviorUpdate$ = genin$.pipe(
    map((s) => s),
    distinctUntilKeyChanged("behavior", (b1, b2) => b1.type === b2.type),
    switchMap((shinobi) => {
      switch (shinobi.behavior.type) {
        case "available": {
          return state$.pipe(
            map((state) => Boolean(IR.list(state.village.tasks).length)),
            distinctUntilChanged(),
            switchMap((going) => (going ? of(assignRandomTask) : EMPTY))
          );
        }
        case "injured": {
          return timer(SHINOBI_INJURED_TIMEOUT_BASE).pipe(map(() => turnIdle));
        }
        case "idle": {
          return timer(SHINOBI_IDLE_TASK_TIMEOUT_BASE).pipe(
            map(() => turnAvailable)
          );
        }
        case "assigned-task":
          return timer(SHINOBI_ASSIGNED_TASK_TIMEOUT_BASE).pipe(
            withLatestFrom(genin$),
            map(([, genin]) => {
              assert(genin.behavior.type === "assigned-task");
              const p = Genin.taskSuccessChance(genin.behavior.task, genin);
              const success = chance(p);
              return success
                ? taskSuccessFn(genin.behavior.task)
                : taskFailureFn(genin.behavior.task);
            })
          );
      }
    })
  );

  return shinobiBehaviorUpdate$;
}
