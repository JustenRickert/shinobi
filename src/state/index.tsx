import { createContext } from "preact";
import {
  ReactNode,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "preact/compat";
import {
  EMPTY,
  Observable,
  Subject,
  distinctUntilChanged,
  from,
  identity,
  map,
  mergeAll,
  tap,
} from "rxjs";
import { IR, shallowEqual } from "../util";

import * as Genin from "./genin";
import * as Task from "./task";
import * as ShinobiInTraining from "./shinobi-in-training";

export { Genin, Task, ShinobiInTraining };

export interface GameState {
  points: number;
  pointsSpent: number;
  genin: {
    ids: Genin.Id[];
    record: Record<Genin.Id, Genin.T>;
  };
  shinobiInTraining: IR.T<ShinobiInTraining.Id, ShinobiInTraining.T>;
  village: {
    tasks: {
      ids: Task.Id[];
      record: Record<Task.Id, Task.T>;
    };
  };
}

const GameContext = createContext<{
  state$: Observable<GameState>;
  update: (reducer: (state: GameState) => GameState) => void;
}>({
  state$: EMPTY,
  update: () => {},
});

let __game_state__: GameState = {
  points: 100,
  pointsSpent: 0,
  genin: {
    ids: [],
    record: {},
  },
  shinobiInTraining: IR.add(
    ShinobiInTraining.make(),
    IR.make<ShinobiInTraining.Id, ShinobiInTraining.T>()
  ),
  village: {
    tasks: {
      ids: [],
      record: {},
    },
  },
};

const gameSubject = new Subject<GameState>();

function getGameState() {
  return __game_state__;
}

function setGameState(state: GameState) {
  __game_state__ = state;
  gameSubject.next(state);
}

export type GameEffect = (
  state$: Observable<GameState>
) => Observable<void | ((state: GameState) => void | GameState)>;

export function runGameEffects(effects: GameEffect[]) {
  const update$ = from(
    effects.map((effect) => effect(gameSubject.asObservable()))
  ).pipe(mergeAll());
  return update$.subscribe((reducer) => {
    if (!reducer) return;
    const nextState = reducer(getGameState());
    if (!nextState) return;
    setGameState(nextState);
  });
}

export function GameProvider({ children }: { children: ReactNode }) {
  const value = useMemo(
    () => ({
      state$: gameSubject.asObservable(),
      update: (reducer: (state: GameState) => GameState) =>
        setGameState(reducer(getGameState())),
    }),
    []
  );
  return <GameContext.Provider value={value}>{children}</GameContext.Provider>;
}

export function useSetGameState() {
  return useContext(GameContext).update;
}

export function useGameStateStream() {
  const { state$ } = useContext(GameContext);
  return state$;
}

export function useGameState<T>(getter: (state: GameState) => T, debug?: any) {
  const state$ = useGameStateStream();
  const [local, setLocal] = useState(() => getter(getGameState()));

  useEffect(() => {
    const sub = state$
      .pipe(
        map(getter),
        distinctUntilChanged((s1, s2) => shallowEqual(s1, s2)),
        debug ? tap(() => console.log(debug)) : identity
      )
      .subscribe((s) => setLocal(s));
    return () => {
      sub.unsubscribe();
    };
  }, [state$]);

  return local;
}

// TODO Having this is nice for dev server, but doesn't work with jest. (Very
// annoying.) Might be possible to work by not using type="module" in
// `package.json` I dunno. Not super important maybe
if (import.meta.hot) {
  import.meta.hot.accept(() => {
    console.log("HIT THAT SUH");
    // better way? Probably not.
    window.location.reload();
  });
}
