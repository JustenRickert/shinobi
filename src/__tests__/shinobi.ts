// TODO fix so we can have unit tests...
import { Shinobi, Task } from "../state/module";

jest.mock("../util", () => ({
  ...jest.requireActual("../util"),
  deviation: (n: number, _p: number) => n,
}));

function cartesian<T>(a: T[], b: T[]): [T, T][];
function cartesian<T>(a: T[], b: T[], c: T[]): [T, T, T][];
function cartesian<T>(...a: T[][]) {
  // @ts-expect-error
  return a.reduce((a, b) => a.flatMap((d) => b.map((e) => [d, e].flat())));
}

describe("shinobi", () => {
  it("task experience scales", () => {
    cartesian([0, 1, 2], [0, 1, 2])
      .map(([shinobiLevel, taskLevel]) => ({
        shinobi: Shinobi.make({ level: shinobiLevel }),
        task: Task.make({ level: taskLevel }),
      }))
      .forEach(({ shinobi, task }) => {
        const xp = Shinobi.taskExperienceGain(task, shinobi);
        expect(
          `shinobi=${shinobi.level}, task=${task.level}, xp=${xp}`
        ).toMatchSnapshot();
      });
  });
});
