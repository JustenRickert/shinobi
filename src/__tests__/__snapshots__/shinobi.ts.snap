// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shinobi task experience scales 1`] = `"shinobi=0, task=0, xp=5"`;

exports[`shinobi task experience scales 2`] = `"shinobi=0, task=1, xp=12"`;

exports[`shinobi task experience scales 3`] = `"shinobi=0, task=2, xp=18"`;

exports[`shinobi task experience scales 4`] = `"shinobi=1, task=0, xp=5"`;

exports[`shinobi task experience scales 5`] = `"shinobi=1, task=1, xp=12"`;

exports[`shinobi task experience scales 6`] = `"shinobi=1, task=2, xp=18"`;

exports[`shinobi task experience scales 7`] = `"shinobi=2, task=0, xp=4"`;

exports[`shinobi task experience scales 8`] = `"shinobi=2, task=1, xp=10"`;

exports[`shinobi task experience scales 9`] = `"shinobi=2, task=2, xp=18"`;
